### PR TITLE
State expire: implementation of MPT R&W

### PR DESCRIPTION
**Pull request by cryyl**
### Description

draft an implement of read and write of MPT and shadow nodes.

### Rationale

Since node may be expired, the MPT Read and write should have related changes.

### Example

None

### Changes

Notable changes: 
* add each change in a bullet point here
* ...

### CODE REVIEW

File: trie/trie.go [hunk 1]
```diff
@@ -355,7 +468,8 @@ func (t *Trie) insert(n node, prefix, key []byte, value node) (bool, node, error
 		if err != nil {
 			return false, nil, err
 		}
-		dirty, nn, err := t.insert(rn, prefix, key, value)
+		rn.setEpoch(epoch)
```
Comment (0xbundler) [inline trie/trie.go:471]: may setEpoch(t.currentEpoch)

File: trie/trie.go [hunk 2]
```diff
@@ -495,7 +637,8 @@ func (t *Trie) delete(n node, prefix, key []byte) (bool, node, error) {
 		if err != nil {
 			return false, nil, err
 		}
-		dirty, nn, err := t.delete(rn, prefix, key)
+		rn.setEpoch(epoch)
```
Comment (0xbundler) [inline trie/trie.go:640]: may setEpoch(t.currentEpoch)?

File: trie/trie.go [hunk 3]
```diff
@@ -355,7 +468,8 @@ func (t *Trie) insert(n node, prefix, key []byte, value node) (bool, node, error
 		if err != nil {
```
Comment (0xbundler) [inline trie/trie.go:468]: where init trie node's shadow nodes?

File: trie/trie.go [hunk 4]
```diff
@@ -166,7 +242,8 @@ func (t *Trie) tryGet(origNode node, key []byte, pos int, epoch uint16) (value [
...
 			return nil, n, true, err
 		}
-		value, newnode, _, err := t.tryGet(child, key, pos, epoch)
+		child.setEpoch(epoch)
```
[the remark this answers is not in the document]
Comment (cryyl) [reply trie/trie.go:245]: this not update epoch, this is set the next epoch, hash node resolve to a shortnode or a fullnode whose epoch is set from its parent.

File: trie/trie.go (hunk 3 above)
Comment (cryyl) [reply trie/trie.go:468]: ChildExpired function will call function GetShadownode which can load shadow node from cache or disk.

File: trie/trie.go (hunk 1 above)
Comment (cryyl) [reply trie/trie.go:471]: this is not update, this is set the next node's epoch index

File: trie/trie.go (hunk 2 above)
Comment (cryyl) [reply trie/trie.go:640]: same as the up one

File: trie/trie.go (hunk 3 above)
Comment (0xbundler) [reply trie/trie.go:468]: may could use trie.sndb to fetch shadow nodes? full node cannot handle shadow nodes query

**Review by 0xbundler — APPROVED** [review]

**Pull request merged**